libcodechecker/server/api/report_server.py

```diff
@@ -1939,66 +1953,153 @@ def massStoreRun(self, name, tag, version, b64zip, force):
                                                    max_run_count,
                                                    remove_run_count))
 
-        with util.TemporaryDirectory() as zip_dir:
-            # Unzip sent data.
-            unzip(b64zip, zip_dir)
+            # If the run can be stored, we need to lock it first.
+            LOG.error("Checking lock for run " + name)
+            lock_record = session.query(RunLock) \
+                .filter(RunLock.name == name) \
+                .with_for_update(nowait=True).one_or_none()
+
+            if not lock_record:
+                # If there is no lock record for the given run name, the run
+                # is not locked -- create a new lock.
+                lock_record = RunLock(name)
+                LOG.critical("Locking brand new now...")
+                session.add(lock_record)
+            elif lock_record.has_expired(
+                    db_cleanup.RUN_LOCK_TIMEOUT_IN_DATABASE):
+                # There can be a lock in the database, which has already
+                # expired. In this case, we assume that the previous operation
+                # has failed, and thus, we can re-use the already present lock.
+                lock_record.touch()
+                LOG.critical("Reusing existing lock")
+            else:
+                # In case the lock exists and it has not expired, we must
+                # consider the run a locked one.
+                LOG.critical("STILL LOCKED!")
+                when = lock_record.when_expires(
+                    db_cleanup.RUN_LOCK_TIMEOUT_IN_DATABASE)
+
+                LOG.info("Refusing to store into run '{0}' as it is locked. "
+                         "Lock will expire at '{1}'.".format(name, when))
+                raise shared.ttypes.RequestFailed(
+                    shared.ttypes.ErrorCode.DATABASE,
+                    "The run named '{0}' is being stored into by another "
+                    "user. If the other store operation has failed, this "
+                    "lock will expire at '{1}'.".format(name, when))
+
+            # At any rate, if the lock has been created or updated, commit it
+            # into the database.
+            LOG.info("committing locks...")
 
-            LOG.debug("Using unzipped folder '{0}'".format(zip_dir))
+            try:
+                session.commit()
+                LOG.info("Locks away.")
+            except sqlalchemy.exc.IntegrityError:
```

Comment:
If there is an existing lock on a run and multiple store try to update this run on the same time I got the following error message:

UPDATE statement on table 'run_locks' expected to update 1 row(s); 0 were matched.
[2018-01-29 12:36] - <class 'sqlalchemy.orm.exc.StaleDataError'>

To reproduce this error by put the following line after this: https://github.com/whisperity/CodeChecker/blob/511a1f9daf29abcb480d0bde4fc2d09c2f34e338/libcodechecker/server/api/report_server.py#L1974
```py
lock_record.touch()
time.sleep(int(tag))
LOG.critical("Reusing existing lock")
```